### PR TITLE
[ME-6] : Implemented the Provider Newness attribute processer

### PR DESCRIPTION
**Pull request by anuragsikakollu**

### CODE REVIEW

File: src/main/java/com/dmg/matchingengine/common/enums/AttributeBucket.java [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package com.dmg.matchingengine.common.enums;
+
+public enum AttributeBucket {
```
Comment (rais-dmg) [inline src/main/java/com/dmg/matchingengine/common/enums/AttributeBucket.java:3]: this is already there?

File: src/main/java/com/dmg/matchingengine/common/enums/AttributeProcessorEnum.java [hunk 2]
```diff
@@ -0,0 +1,6 @@
+package com.dmg.matchingengine.common.enums;
+
+public enum AttributeProcessorEnum {
+
```
Comment (rais-dmg) [inline src/main/java/com/dmg/matchingengine/common/enums/AttributeProcessorEnum.java:4]: moved to enum package?

File: src/main/java/com/dmg/matchingengine/model/entity/ProviderEntity.java [hunk 3]
```diff
@@ -28,4 +29,7 @@
 
     @Column(name = "open_job_application_count", nullable = false)
     private int openJobApplicationCount;
+
+    @Column(name = "completed_onboarding_utc", columnDefinition = "TIMESTAMP")
```
Comment (rais-dmg) [inline src/main/java/com/dmg/matchingengine/model/entity/ProviderEntity.java:33]: yu can also use     @Temporal(TemporalType.TIMESTAMP)

File: src/main/java/com/dmg/matchingengine/service/NewnessAttributeProcessor.java [hunk 4]
```diff
@@ -1,30 +1,76 @@
...
+
+    private AttributeBucket compute(String providerId, WorkType workType) {
+
+        ProviderEntity provider = providerService.get(providerId);
+        WorkTypeConfig.AttributeProcessorConfig config = getProcessorConfig(workType);
+
+        long durationSinceOnboarding =  ChronoUnit.valueOf(config.getProperties().get(WORK_TYPE_CONFIG_PROPERTY_UNIT))
+            .between(provider.getCompletedOnboardingUtc(), LocalDateTime.now());
+
+        if(durationSinceOnboarding <= Long.parseLong(config.getBucket().get(AttributeBucket.HIGH)))
```
Comment (rais-dmg) [inline src/main/java/com/dmg/matchingengine/service/NewnessAttributeProcessor.java:63]: a better way is to fetch attribute enum values and iterate through them

File: src/main/java/com/dmg/matchingengine/service/OpenJobsAttributeProcessor.java [hunk 5]
```diff
@@ -61,21 +62,21 @@ private AttributeBucketEnum compute(String providerId, WorkType workType) {
             });
 
 
-        AttributeProviderProperties properties = getProcessorConfig(workType);
+        AttributeProcessorConfig properties = getProcessorConfig(workType);
 
         if (provider.getOpenJobCount() > Integer.parseInt(
-            properties.getBucket().get(AttributeBucketEnum.HIGH))) {
-            return AttributeBucketEnum.HIGH;
+            properties.getBucket().get(AttributeBucket.HIGH))) {
```
Comment (rais-dmg) [inline src/main/java/com/dmg/matchingengine/service/OpenJobsAttributeProcessor.java:68]: iterate through enums

File: src/main/java/com/dmg/matchingengine/common/enums/AttributeBucket.java (hunk 1 above)
Comment (rais-dmg) [reply src/main/java/com/dmg/matchingengine/common/enums/AttributeBucket.java:3]: can yu please modify these enums order as High, medium , low. Order is important for us.

File: src/main/java/com/dmg/matchingengine/service/ProviderService.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+@Service
+public class ProviderService {
+
+    private final ProviderRepository providerRepository;
+
+    public ProviderService(final ProviderRepository providerRepository) {
+        this.providerRepository = providerRepository;
+    }
+
+    public ProviderEntity get(String providerId) {
```
Comment (dmittal-dmg) [inline src/main/java/com/dmg/matchingengine/service/ProviderService.java:22]: name it getEntityById or something.

File: src/main/java/com/dmg/matchingengine/common/enums/AttributeProcessorEnum.java (hunk 2 above)
Comment (anuragsikakollu) [reply src/main/java/com/dmg/matchingengine/common/enums/AttributeProcessorEnum.java:4]: There is another class with the name `AttributeProcessor` so had to keep the `Enum` suffix

File: src/main/java/com/dmg/matchingengine/service/ScoreService.java [hunk 7]
```diff
@@ -33,9 +34,9 @@
     private final ThreadPoolExecutor executor;
 
     public ScoreService(
-        Set<AttributeProcessor> attributeProcessors) {
+        Set<com.dmg.matchingengine.service.AttributeProcessor> attributeProcessors) {
```
[the remark this answers is not in the document]
Comment (anuragsikakollu) [reply src/main/java/com/dmg/matchingengine/service/ScoreService.java:37]: corrected this

File: src/main/java/com/dmg/matchingengine/service/ProviderService.java (hunk 6 above)
Comment (anuragsikakollu) [reply src/main/java/com/dmg/matchingengine/service/ProviderService.java:22]: done

File: src/main/java/com/dmg/matchingengine/service/OpenJobsAttributeProcessor.java (hunk 5 above)
Comment (anuragsikakollu) [reply src/main/java/com/dmg/matchingengine/service/OpenJobsAttributeProcessor.java:68]: I prefer this, as it is very simple, readable, less complex, very deubug-able and developer friendly. 
Especially  it does not depend on the order of `AttributeBucket ` enum, thus less scope of error

File: src/main/java/com/dmg/matchingengine/common/enums/AttributeBucket.java (hunk 1 above)
Comment (anuragsikakollu) [reply src/main/java/com/dmg/matchingengine/common/enums/AttributeBucket.java:3]: The order of enums should not matter, if code depends on the order of enums in this file, then there is something wrong with the code as something as simple as the order can mess up the entire code

**Review by dmittal-dmg — APPROVED** [review]

**Pull request merged**